Add module-info.java to support JPMS

#### Describe the feature

The `vertx-launcher-application` jar should contain a `module-info.class` file, making it a named / explicit module. 

#### Use cases

As part of the Vert.x 5.0 release Vert.x tries to support the Java Module System, see [What's new in Vert.x 5, Java Modules](https://vertx.io/blog/whats-new-in-vert-x-5/#java-modules).
By providing a `module-info.java` the VertxApplication class would be usable as a launcher with `jlink`. Building a runtime image with `jlink` is only supported with named / explicit modules.

##### Example without `VertxApplication`

By starting the module through a main method in the client code itself it's currently possible to use `jlink`.

```java
public class MainVerticle extends VerticleBase {

	public static void main(String[] args) {
		Vertx vertx = Vertx.vertx();
		vertx.deployVerticle(new MainVerticle())
			.onFailure(Throwable::printStackTrace);
	}

  @Override
  public Future<?> start() {
    return vertx.createHttpServer().requestHandler(req -> {
      req.response()
        .putHeader("content-type", "text/plain")
        .end("Hello from Vert.x!");
    }).listen(8888).onSuccess(http -> {
      System.out.println("HTTP server started on port 8888");
    });
  }
}
```

```java
module com.example.module_example {
	requires io.vertx.core;
}
```

```kotlin
val launcherModule = "com.example.module_example"
val launcherClassName = "com.example.module_example.MainVerticle"

application {
  mainModule.set(launcherModule)
  mainClass.set(launcherClassName)
}

tasks.register<JavaExec>("jlink") {
	group = "build"

    mainClass.set("jdk.tools.jlink.internal.Main")
    mainModule.set("jdk.jlink")

    var modulePath = files()
    modulePath.from(configurations.runtimeClasspath)
    modulePath.from(tasks.jar)
    println(modulePath.joinToString(":"))
    args = listOf(
            "--module-path",
            modulePath.joinToString(":"),
            "--add-modules",
            "java.base,java.compiler,java.logging,java.naming,jdk.unsupported,${application.mainModule.get()}",
            "--output",
            "build/jlink-image-test",
            "--launcher",
            "launch=${application.mainModule.get()}/${application.mainClass.get()}"
    )
}
```

##### Currently not working

```java
public class MainVerticle extends VerticleBase {

  @Override
  public Future<?> start() {
    return vertx.createHttpServer().requestHandler(req -> {
      req.response()
        .putHeader("content-type", "text/plain")
        .end("Hello from Vert.x!");
    }).listen(8888).onSuccess(http -> {
      System.out.println("HTTP server started on port 8888");
    });
  }
}
```

```java
module com.example.module_example {
	requires io.vertx.core;

	exports com.example.module_example to vertx.launcher.application, io.vertx.core;
}
```

```kotlin
val launcherModule = "vertx.launcher.application"
val launcherClassName = "io.vertx.launcher.application.VertxApplication"

application {
  mainModule.set(launcherModule)
  mainClass.set(launcherClassName)
}

tasks.register<JavaExec>("jlink") {
	group = "build"

    mainClass.set("jdk.tools.jlink.internal.Main")
    mainModule.set("jdk.jlink")

    var modulePath = files()
    modulePath.from(configurations.runtimeClasspath)
    modulePath.from(tasks.jar)
    println(modulePath.joinToString(":"))
    args = listOf(
            "--module-path",
            modulePath.joinToString(":"),
            "--add-modules",
            "java.base,java.compiler,java.logging,java.naming,jdk.unsupported,${application.mainModule.get()}",
            "--output",
            "build/jlink-image-test",
            "--launcher",
            "launch=${application.mainModule.get()}/${application.mainClass.get()}"
    )
}
```

#### Contribution

Who should implement this feature ? are you volunteering for implementing this feature or
do you know that is able and willing implement this feature ?

I think this would do the trick:

```java
module vertx.application.launcher {
	requires io.vertx.core;
	requires io.vertx.core.logging;
	requires io.vertx.docgen;

	requires info.picocli;
}
```

But the launcher needs one more package that `vertx.core` currently [doesn't export to it](https://github.com/eclipse-vertx/vert.x/blob/3d281cb2b05c26c209111e32478c93639183f0ed/vertx-core/src/main/java/module-info.java#L98):

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.13.0:compile (default-compile) on project vertx-launcher-application: Compilation failure
[ERROR] vertx-launcher/application/src/main/java/io/vertx/launcher/application/impl/VertxApplicationCommand.java:[16,21] package io.vertx.core.impl is not visible
[ERROR]   (package io.vertx.core.impl is declared in module io.vertx.core, which does not export it to module vertx.application.launcher)
```

--- a/application/pom.xml
+++ b/application/pom.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
-  ~ Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+  ~ Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
   ~
   ~ This program and the accompanying materials are made available under the
   ~ terms of the Eclipse Public License 2.0 which is available at
@@ -86,6 +86,19 @@
       <artifactId>log4j-core</artifactId>
       <scope>test</scope>
     </dependency>
+    <!-- Required for JPMS integration testing -->
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.13.2</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.hamcrest</groupId>
+      <artifactId>hamcrest-core</artifactId>
+      <version>3.0</version>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <build>
@@ -119,6 +132,32 @@
       </plugins>
     </pluginManagement>
     <plugins>
+      <plugin>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <configuration>
+          <useModulePath>false</useModulePath>
+          <excludes>
+            <exclude>**/*ITest</exclude>
+          </excludes>
+        </configuration>
+      </plugin>
+      <plugin>
+        <artifactId>maven-failsafe-plugin</artifactId>
+        <configuration>
+          <useModulePath>true</useModulePath>
+          <includes>
+            <include>**/*ITest</include>
+          </includes>
+        </configuration>
+        <executions>
+          <execution>
+            <goals>
+              <goal>integration-test</goal>
+              <goal>verify</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
       <plugin>
         <artifactId>maven-assembly-plugin</artifactId>
         <executions>

--- a/application/src/main/java/io/vertx/launcher/application/impl/VertxApplicationCommand.java
+++ b/application/src/main/java/io/vertx/launcher/application/impl/VertxApplicationCommand.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -13,7 +13,6 @@ package io.vertx.launcher.application.impl;
 
 import io.vertx.core.*;
 import io.vertx.core.eventbus.EventBusOptions;
-import io.vertx.core.impl.ServiceHelper;
 import io.vertx.core.internal.VertxInternal;
 import io.vertx.core.internal.logging.Logger;
 import io.vertx.core.json.JsonObject;
@@ -30,7 +29,9 @@ import picocli.CommandLine.Option;
 import picocli.CommandLine.Parameters;
 
 import java.time.Duration;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.ServiceLoader;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeoutException;
@@ -39,7 +40,6 @@ import java.util.function.Supplier;
 import static io.vertx.launcher.application.impl.Utils.*;
 import static java.lang.Boolean.TRUE;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
-import static java.util.stream.Collectors.toList;
 import static picocli.CommandLine.Parameters.NULL_VALUE;
 
 @Command(name = "VertxApplication", description = "Runs a Vert.x application.", sortOptions = false)
@@ -279,11 +279,13 @@ public class VertxApplicationCommand implements Runnable {
   }
 
   private <SP> SP findServiceProvider(Class<SP> serviceProviderClass) {
-    List<SP> serviceProviders = ServiceHelper.loadFactories(VertxServiceProvider.class).stream()
-      .filter(vertxServiceProvider -> serviceProviderClass.isAssignableFrom(vertxServiceProvider.getClass()))
-      .map(serviceProviderClass::cast)
-      .collect(toList());
-    if (serviceProviders.size() == 0) {
+    List<SP> serviceProviders = new ArrayList<>();
+    for (VertxServiceProvider vsp : ServiceLoader.load(VertxServiceProvider.class)) {
+      if (serviceProviderClass.isAssignableFrom(vsp.getClass())) {
+        serviceProviders.add(serviceProviderClass.cast(vsp));
+      }
+    }
+    if (serviceProviders.isEmpty()) {
       return null;
     }
     if (serviceProviders.size() != 1) {

--- a/application/src/main/java/module-info.java
+++ b/application/src/main/java/module-info.java
@@ -1,0 +1,24 @@
+/*
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+module io.vertx.launcher.application {
+  requires info.picocli;
+  requires io.vertx.core;
+  requires io.vertx.core.logging;
+
+  // Annotation processing
+  requires static io.vertx.docgen;
+
+  uses io.vertx.core.spi.VertxServiceProvider;
+  opens io.vertx.launcher.application.impl to info.picocli;
+
+  exports io.vertx.launcher.application;
+}

--- a/application/src/test/java/io/vertx/launcher/application/tests/CustomApplicationLowMemoryTest.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/CustomApplicationLowMemoryTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -9,9 +9,12 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  */
 
-package io.vertx.launcher.application;
+package io.vertx.launcher.application.tests;
 
 import io.vertx.core.AbstractVerticle;
+import io.vertx.launcher.application.HookContext;
+import io.vertx.launcher.application.VertxApplication;
+import io.vertx.launcher.application.VertxApplicationHooks;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;

--- a/application/src/test/java/io/vertx/launcher/application/tests/CustomMetricsFactory.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/CustomMetricsFactory.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -9,8 +9,9 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  */
 
-package io.vertx.core;
+package io.vertx.launcher.application.tests;
 
+import io.vertx.core.VertxOptions;
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.metrics.MetricsOptions;
 import io.vertx.core.spi.VertxMetricsFactory;

--- a/application/src/test/java/io/vertx/launcher/application/tests/CustomMetricsOptions.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/CustomMetricsOptions.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -9,7 +9,7 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  */
 
-package io.vertx.core;
+package io.vertx.launcher.application.tests;
 
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.metrics.MetricsOptions;

--- a/application/src/test/java/io/vertx/launcher/application/tests/HttpTestVerticle.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/HttpTestVerticle.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -9,7 +9,7 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  */
 
-package io.vertx.launcher.application;
+package io.vertx.launcher.application.tests;
 
 import io.vertx.core.AbstractVerticle;
 import io.vertx.core.json.JsonObject;

--- a/application/src/test/java/io/vertx/launcher/application/tests/ModularityITest.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/ModularityITest.java
@@ -1,0 +1,36 @@
+/*
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.launcher.application.tests;
+
+import io.vertx.launcher.application.VertxApplication;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class ModularityITest {
+
+  private final TestHooks hooks = new TestHooks();
+
+  @AfterEach
+  public void tearDown() {
+    if (hooks.vertx != null) {
+      hooks.vertx.close().await();
+    }
+  }
+
+  @Test
+  void shouldDeployVerticleOfModule() {
+    VertxApplication vertxApplication = new VertxApplication(new String[]{"io.vertx.launcher.application.tests.HttpTestVerticle"}, hooks);
+    assertEquals(0, vertxApplication.launch());
+  }
+}

--- a/application/src/test/java/io/vertx/launcher/application/tests/NestedMetricsOptions.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/NestedMetricsOptions.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -9,7 +9,7 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  */
 
-package io.vertx.core;
+package io.vertx.launcher.application.tests;
 
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.metrics.MetricsOptions;

--- a/application/src/test/java/io/vertx/launcher/application/tests/SystemExitTest.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/SystemExitTest.java
@@ -1,6 +1,19 @@
-package io.vertx.launcher.application;
+/*
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
 
-import io.vertx.test.TestVerticle;
+package io.vertx.launcher.application.tests;
+
+import io.vertx.launcher.application.HookContext;
+import io.vertx.launcher.application.VertxApplication;
+import io.vertx.launcher.application.VertxApplicationHooks;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;

--- a/application/src/test/java/io/vertx/launcher/application/tests/TestHooks.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/TestHooks.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -9,9 +9,11 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  */
 
-package io.vertx.launcher.application;
+package io.vertx.launcher.application.tests;
 
 import io.vertx.core.Vertx;
+import io.vertx.launcher.application.HookContext;
+import io.vertx.launcher.application.VertxApplicationHooks;
 
 public class TestHooks implements VertxApplicationHooks {
 

--- a/application/src/test/java/io/vertx/launcher/application/tests/TestVerticle.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/TestVerticle.java
@@ -1,4 +1,15 @@
-package io.vertx.test;
+/*
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.launcher.application.tests;
 
 import io.vertx.core.AbstractVerticle;
 import io.vertx.core.json.JsonObject;

--- a/application/src/test/java/io/vertx/launcher/application/tests/VertxApplicationExtensibilityTest.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/VertxApplicationExtensibilityTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -9,7 +9,7 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  */
 
-package io.vertx.launcher.application;
+package io.vertx.launcher.application.tests;
 
 import io.vertx.core.Verticle;
 import io.vertx.core.VertxBuilder;
@@ -17,6 +17,9 @@ import io.vertx.core.VertxOptions;
 import io.vertx.core.internal.VertxInternal;
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.spi.metrics.VertxMetrics;
+import io.vertx.launcher.application.HookContext;
+import io.vertx.launcher.application.VertxApplication;
+import io.vertx.launcher.application.VertxApplicationHooks;
 import io.vertx.test.fakecluster.FakeClusterManager;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Test;
@@ -24,8 +27,8 @@ import org.junit.jupiter.api.Test;
 import java.io.IOException;
 import java.util.function.Supplier;
 
-import static io.vertx.launcher.application.VertxApplicationTest.assertServerStarted;
-import static io.vertx.launcher.application.VertxApplicationTest.getContent;
+import static io.vertx.launcher.application.tests.VertxApplicationTest.assertServerStarted;
+import static io.vertx.launcher.application.tests.VertxApplicationTest.getContent;
 import static java.lang.Boolean.TRUE;
 import static org.junit.jupiter.api.Assertions.*;
 
@@ -38,7 +41,6 @@ public class VertxApplicationExtensibilityTest {
     if (hooks != null && hooks.vertx != null) {
       hooks.vertx.close();
     }
-    FakeClusterManager.reset();
   }
 
   @Test

--- a/application/src/test/java/io/vertx/launcher/application/tests/VertxApplicationTest.java
+++ b/application/src/test/java/io/vertx/launcher/application/tests/VertxApplicationTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -9,17 +9,17 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
  */
 
-package io.vertx.launcher.application;
+package io.vertx.launcher.application.tests;
 
-import io.vertx.core.CustomMetricsOptions;
 import io.vertx.core.Vertx;
 import io.vertx.core.VertxOptions;
 import io.vertx.core.internal.VertxInternal;
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.metrics.MetricsOptions;
 import io.vertx.core.spi.VertxServiceProvider;
-import io.vertx.test.TestVerticle;
-import io.vertx.test.fakecluster.FakeClusterManager;
+import io.vertx.launcher.application.HookContext;
+import io.vertx.launcher.application.VertxApplication;
+import io.vertx.launcher.application.VertxApplicationHooks;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -78,7 +78,6 @@ public class VertxApplicationTest {
         .atMost(Duration.ofSeconds(10))
         .until(() -> future.isDone());
     }
-    FakeClusterManager.reset();
   }
 
   private void setManifest(String name) throws Exception {
@@ -234,7 +233,7 @@ public class VertxApplicationTest {
       }
     };
     ClassLoader oldCL = Thread.currentThread().getContextClassLoader();
-    Thread.currentThread().setContextClassLoader(createMetricsFromMetaInfLoader("io.vertx.core.CustomMetricsFactory"));
+    Thread.currentThread().setContextClassLoader(createMetricsFromMetaInfLoader("io.vertx.launcher.application.tests.CustomMetricsFactory"));
     try {
       System.setProperty("vertx.metrics.options.enabled", "true");
       TestVertxApplication app = new TestVertxApplication(new String[0], hooks);
@@ -345,7 +344,7 @@ public class VertxApplicationTest {
     };
 
     ClassLoader oldCL = Thread.currentThread().getContextClassLoader();
-    Thread.currentThread().setContextClassLoader(createMetricsFromMetaInfLoader("io.vertx.core.CustomMetricsFactory"));
+    Thread.currentThread().setContextClassLoader(createMetricsFromMetaInfLoader("io.vertx.launcher.application.tests.CustomMetricsFactory"));
     try {
       System.setProperty("vertx.options.eventLoopPoolSize", "123");
       System.setProperty("vertx.options.maxEventLoopExecuteTime", "123767667");
@@ -393,7 +392,7 @@ public class VertxApplicationTest {
       }
     };
     ClassLoader oldCL = Thread.currentThread().getContextClassLoader();
-    Thread.currentThread().setContextClassLoader(createMetricsFromMetaInfLoader("io.vertx.core.CustomMetricsFactory"));
+    Thread.currentThread().setContextClassLoader(createMetricsFromMetaInfLoader("io.vertx.launcher.application.tests.CustomMetricsFactory"));
     try {
       System.setProperty("vertx.metrics.options.enabled", "true");
       System.setProperty("vertx.metrics.options.customProperty", "customPropertyValue");

--- a/application/src/test/java/module-info.java
+++ b/application/src/test/java/module-info.java
@@ -1,0 +1,21 @@
+/*
+ * Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+open module io.vertx.launcher.application.tests {
+  requires io.vertx.core;
+  requires io.vertx.launcher.application;
+
+  requires org.junit.jupiter.api;
+
+  // Only required for compilation
+  requires static io.vertx.core.tests;
+  requires static awaitility;
+}

--- a/application/src/test/resources/META-INF/MANIFEST-Http-Verticle.MF
+++ b/application/src/test/resources/META-INF/MANIFEST-Http-Verticle.MF
@@ -1,2 +1,2 @@
-Main-Class: io.vertx.launcher.application.VertxApplicationTest$TestVertxApplication
-Main-Verticle: io.vertx.launcher.application.HttpTestVerticle
+Main-Class: io.vertx.launcher.application.tests.VertxApplicationTest$TestVertxApplication
+Main-Verticle: io.vertx.launcher.application.tests.HttpTestVerticle

--- a/application/src/test/resources/META-INF/MANIFEST-Missing-Main-Verticle.MF
+++ b/application/src/test/resources/META-INF/MANIFEST-Missing-Main-Verticle.MF
@@ -1,2 +1,2 @@
-Main-Class: io.vertx.launcher.application.VertxApplicationTest$TestVertxApplication
+Main-Class: io.vertx.launcher.application.tests.VertxApplicationTest$TestVertxApplication
 Main-Verticle: org.acme.MissingVerticle

--- a/application/src/test/resources/META-INF/services/io.vertx.core.spi.VertxServiceProvider
+++ b/application/src/test/resources/META-INF/services/io.vertx.core.spi.VertxServiceProvider
@@ -1,1 +1,12 @@
-io.vertx.core.CustomMetricsFactory
+#
+# Copyright (c) 2011-2025 Contributors to the Eclipse Foundation
+#
+# This program and the accompanying materials are made available under the
+# terms of the Eclipse Public License 2.0 which is available at
+# http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+# which is available at https://www.apache.org/licenses/LICENSE-2.0.
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+#
+
+io.vertx.launcher.application.tests.CustomMetricsFactory

